contracts/cw1-subkeys/src/contract.rs

```diff
@@ -216,14 +216,38 @@ pub fn query_allowance<S: Storage, A: Api, Q: Querier>(
 #[cfg(test)]
 mod tests {
     use super::*;
-    use cosmwasm_std::testing::{mock_dependencies, mock_env, MOCK_CONTRACT_ADDR};
-    use cosmwasm_std::{coins, BankMsg, StdError, WasmMsg};
+    use crate::balance::Balance;
+    use cosmwasm_std::testing::{mock_dependencies, mock_env};
+    use cosmwasm_std::{coin, coins};
 
-    const CANONICAL_LENGTH: usize = 20;
+    // this will set up the init for other tests
+    fn setup_test_case<S: Storage, A: Api, Q: Querier>(
+        mut deps: &mut Extern<S, A, Q>,
+        env: &Env,
+        admins: &Vec<HumanAddr>,
+        spenders: &Vec<HumanAddr>,
+        allowances: &Vec<Coin>,
+        expirations: &Vec<Expiration>,
+    ) {
+        // Init a contract with admins
+        let init_msg = InitMsg {
+            admins: admins.clone(),
+            mutable: false,
+        };
+        init(deps, env.clone(), init_msg).unwrap();
 
-    // you probably want some `setup_test_case` function that inits a contract with 2 admins
-    // and 2 subkeys with some allowances. these keys can be constants strings here,
-    // used like `HumanAddr::from(admin1)` and then all tests can just run against that
+        // Add subkeys with initial allowances
+        for (spender, expiration) in spenders.iter().zip(expirations) {
```

Comment:
interesting zip. works well actually.

I have used `&[(HumanAddr, Expiration)]` in similar cases, which made more sense to me, but also may be harder to construct (if you have a list of admins anyway)

Reply:
Actually reading https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.zip
It seems that the test case below with 4 spenders and 2 expirations, only the first two spenders will be registered as `zip` ends when the first of the two iterators ends.

That would not be my expected behavior from this function signature, and should be modified or documented with rustdoc at the function level

Reply:
> It seems that the test case below with 4 spenders and 2 expirations, only the first two spenders will be registered as `zip` ends when the first of the two iterators ends.

Yes, but I've put only two spenders in the array. Maybe renaming it to `initial_spenders` is a good idea.